contractcourt/channel_arbitrator.go

```diff
@@ -1365,15 +1414,54 @@ func (c *ChannelArbitrator) channelAttendant(bestHeight int32) {
 
 		// We've cooperatively closed the channel, so we're no longer
 		// needed. We'll mark the channel as resolved and exit.
-		case <-c.cfg.ChainEvents.CooperativeClosure:
-			log.Infof("ChannelArbitrator(%v) closing due to co-op "+
-				"closure", c.cfg.ChanPoint)
+		case closeInfo := <-c.cfg.ChainEvents.CooperativeClosure:
+			log.Infof("ChannelArbitrator(%v) marking channel "+
+				"cooperatively closed", c.cfg.ChanPoint)
+
+			// Create a summary and mark the channel cooperatively
+			// closed in the database.
+			closeSummary := &channeldb.ChannelCloseSummary{
+				ChanPoint:      closeInfo.ChannelPoint,
+				ChainHash:      closeInfo.ChainHash,
+				ClosingTXID:    *closeInfo.SpenderTxHash,
+				RemotePub:      &closeInfo.RemoteIdentity,
+				Capacity:       closeInfo.Capacity,
+				CloseHeight:    uint32(closeInfo.SpendingHeight),
+				SettledBalance: closeInfo.SettledBalance,
+				CloseType:      channeldb.CooperativeClose,
+				ShortChanID:    c.cfg.ShortChanID,
+				IsPending:      false,
+			}
+
+			err := c.cfg.MarkChannelClosed(closeSummary)
+			if err != nil {
+				log.Errorf("unable to mark channel closed: "+
+					"%v", err)
+				return
+			}
 
+			// Since all outputs are imemdiately available in a
+			// cooperative close, we can mark the channel resolved
+			// immediately.
 			if err := c.cfg.MarkChannelResolved(); err != nil {
 				log.Errorf("Unable to mark contract "+
 					"resolved: %v", err)
+				return
+			}
+
+			// To clean up, we ensure we commit to StateFullyResolved.
+			c.state = StateFullyResolved
```

Comment:
What is the rationale for reimplementing channel resolved + commit state here, instead of using `advanceState`? The latter seems more fault tolerant as rn idt we will properly recover if either the `Closed -> Resolved` or `Resolved -> State Committed` transitions fail

Reply:
we won't though, because the channel has already been closed. we only resubscribe for open channels.

Reply:
I agree that it would be cleanest to make coop closes also go through the state machine. It would require a new trigger type, and probably also writing (empty) resolutions for the coop close case. Wouldn't be that big of a change though, and would be nice to have all closes (except breaches) follow the same logic :)

Regarding replaying of this state, it is benign, as after the coop close has been called with `CloseChannel` all the necessary state is written in the database, `MarkChannelResolved` will only clean up some in-memory state, and `CommitState` will write the state to the DB, but it will never be read again.

Let me know what you think!

Reply:
After @cfromknecht's suggestion I added a new trigger `coopCloseTrigger` and make the state machine skip ahead to `StateFullyResolved` in case it triggers. Very happy with the result, a lot cleaner now! 👍